Add mutliplayer cookbook recipe

filiph opened the pull request:
_Description of what this PR is changing or adding, and why:_
A cookbook recipe for adding multiplayer to a Flutter game via Firebase's Cloud Firestore.

Marked as `[WIP]` because:

- Illustrations are missing
- The `card_game` template on which this cookbook is based has not yet landed at github.com/flutter/games

But the text is ready for review. (Already edited by @atsansone in its original Google Docs form.)

_Issues fixed by this PR (if any):_
Fixes #9669.

## Presubmit checklist

- [x] This PR doesn’t contain automatically generated corrections (Grammarly or similar).
- [x] This PR follows the [Google Developer Documentation Style Guidelines](https://developers.google.com/style) — for example, it doesn’t use _i.e._ or _e.g._, and it avoids _I_ and _we_ (first person).
- [x] This PR uses [semantic line breaks](https://github.com/dart-lang/site-shared/blob/main/doc/writing-for-dart-and-flutter-websites.md#semantic-line-breaks) of 80 characters or fewer.

domesticmouse commented:
This requires work to bring it into alignment with the new base

filiph commented:
> This requires work to bring it into alignment with the new base

Done.

zoeyfan commented:
@filiph Is this ready to merge?

zoeyfan commented:
The URL will be: https://docs.flutter.dev/cookbook/games/firestore-multiplayer

filiph commented:
Sorry for the delay, but this is now ready. The text is the same as has been reviewed. I added illustrations and made sure the recipe works with the final form of the `flutter/games` repo.

filiph commented:
For reference, here's the full page, including the pictures. I strive to make pages as graphically varied as possible. I think it invites study, as opposed to pages that have large blocks of unbroken text.

![Screenshot 2023-11-09 at 15-38-09 Add multiplayer support via Firestore](https://github.com/flutter/website/assets/919717/4b53a9fb-25de-435f-bca2-2e482817bea8)

Also, @parlough, would you be able to have a quick look, if you haven't already? I think this is the last big one to land. The rest is just "details".

parlough commented on src/cookbook/games/firestore-multiplayer.md:
````diff
@@ -0,0 +1,469 @@
+in the [`flutter/games` repository](https://github.com/flutter/games).
+Run the following command to clone that repository:
+
+```shell
````
````suggestion
```terminal
````

parlough commented on src/cookbook/games/firestore-multiplayer.md:
```diff
@@ -0,0 +1,469 @@
+If you want a quick, 15-minute primer on Cloud Firestore,
+watch the following video:
+
+<iframe width="560" height="315" src="https://www.youtube.com/embed/v_hR4K4auoQ" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>
```
We're starting to include the video title in the iframe `title`:

```suggestion
<iframe width="560" height="315" src="https://www.youtube.com/embed/v_hR4K4auoQ" title="YouTube video player - What is a NoSQL Database? How is Cloud Firestore structured? | Get to know Cloud Firestore #1" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>
```

parlough commented on src/cookbook/games/firestore-multiplayer.md:
```diff
@@ -0,0 +1,469 @@
+
+<iframe width="560" height="315" src="https://www.youtube.com/embed/v_hR4K4auoQ" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>
+
+To add Firestore to your Flutter project,
+follow the first two steps of the
+[Get started with Cloud Firestore](https://firebase.google.com/docs/firestore/quickstart)
+guide.
```
I'm a bit worried that "first two steps" will get out of date and as a reader, it may not be clear if this means the numbered steps within "Create a Cloud Firestore database" or both "Create a Cloud Firestore database" and "Set up your development environment".

Could we perhaps be more explicit somehow? If we directly linked to the related headers, we could catch changes automatically with the linkchecker.

```suggestion
To add Firestore to your Flutter project,
follow the first two steps of the
[Get started with Cloud Firestore](https://firebase.google.com/docs/firestore/quickstart)
guide:

- [Create a Cloud Firestore database](https://firebase.google.com/docs/firestore/quickstart#create)
- [Set up your development environment](https://firebase.google.com/docs/firestore/quickstart#set_up_your_development_environment)

```

parlough commented on src/cookbook/games/firestore-multiplayer.md:
````diff
@@ -0,0 +1,469 @@
+    in `lib/main.dart`:
+
+    ```dart
+     WidgetsFlutterBinding.ensureInitialized();
+    
+      await Firebase.initializeApp(
+        options: DefaultFirebaseOptions.currentPlatform,
+      );
````
```suggestion
     await Firebase.initializeApp(
       options: DefaultFirebaseOptions.currentPlatform,
     );
```

parlough replied:
```suggestion
    WidgetsFlutterBinding.ensureInitialized();
    
    await Firebase.initializeApp(
      options: DefaultFirebaseOptions.currentPlatform,
    );
```

parlough commented on src/cookbook/games/firestore-multiplayer.md:
```diff
@@ -0,0 +1,469 @@
+
+    This ensures that Firebase is initialized on game startup.
+
+3.  Add the Firestore instance to the app.
+    That way, any widget can access this instance.
+    Widgets can also react to the instance missing if needed.
+
+    Replace the usual `runApp(MyApp())` with the following:
```
I feel this step jumps a bit. It's mentioned earlier these steps can be followed without the card template, but this step assumes the user has provider as a dependency and they know what it is/what it's doing.

Could this be expanded on a bit or at least the requirement mentioned (`dart pub add provider`)?

filiph replied:
Good point, I'll address this.

parlough commented on src/cookbook/games/firestore-multiplayer.md:
```diff
@@ -0,0 +1,469 @@
+
+  /// For now, there is only one match. But in order to be ready
+  /// for match-making, put it in a Firestore collection called matches.
+  late final _matchRef = instance.collection('matches').doc('match_1');
+
+  late final _areaOneRef = _matchRef
+      .collection('areas')
+      .doc('area_one')
+      .withConverter<List<PlayingCard>>(
+          fromFirestore: _cardsFromFirestore, toFirestore: _cardsToFirestore);
+
+  late final _areaTwoRef = _matchRef
+      .collection('areas')
+      .doc('area_two')
+      .withConverter<List<PlayingCard>>(
+          fromFirestore: _cardsFromFirestore, toFirestore: _cardsToFirestore);
```
Not 100% necessary, just a question. Would it be useful to add the explicit types here? I generally do it for fields as a best practice, but it also might help readers understand what's being returned since this is on a web page rather than IDE.

filiph replied:
I'm going to take a note and revisit this later. I see your point but I'm torn. The Dart style is not to add extraneous types to private members. On the other hand, this particular piece of code _is_ dense — I can totally see people wanting to have the types explicit there.

domesticmouse replied:
I'd lean against over specifying the types here. Looking to see what the community is doing lead to this S/O post:

https://stackoverflow.com/questions/68079030/how-to-use-firestore-withconverter-in-flutter

I think this code is good as is. =)

parlough replied:
Works for me :)

parlough commented on src/cookbook/games/firestore-multiplayer.md:
```diff
@@ -0,0 +1,469 @@
+    makes it appear on the other one.
+
+    {% comment %}
+      TBA: GIF of multiplayer working
```
Just checking in to see if you plan to add this now or are tracking adding it later.

filiph replied:
This is for later. Maybe just a few days for now, but I don't want to block landing this on it, because making a good gif always takes me quite a lot of time.

parlough reviewed (approved):
Thanks @filiph! I love the illustrations as usual and this recipe reads really well.

I also always love the general discussion you include about development, in this case game development. It makes the recipe a more complete learning experience that helps the reader stay engaged.

One concern I have about the provider step and a few minor suggestions:

filiph commented:
Thanks for the suggestions, @parlough! I've implemented all of them and addressed most of your notes.

I think I'm ready to merge. I just need to stage this and verify everything is as it should be.

parlough commented:
Looks good to me, thanks so much @filiph!
